File: skills/cp-solver/references/implementations/hld/hld.py
"""
Heavy-Light Decomposition (HLD) — Path queries trên cây
Complexity: O(N log N) preprocessing, O(log² N) per query
Dùng khi: path queries với updates, N ≤ 10^5

Đây là skeleton — kết hợp với Segment Tree để query path sum/max/min.
"""
import sys
from collections import defaultdict
sys.setrecursionlimit(300000)
input = sys.stdin.readline


class HLD:
    def __init__(self, n, root=0):
        self.n = n
        self.root = root
        self.adj = defaultdict(list)

        # HLD arrays
        self.parent   = [-1] * n
        self.depth    = [0]  * n
        self.subtree  = [1]  * n
        self.heavy    = [-1] * n   # heavy child của mỗi node
        self.head     = [0]  * n   # đầu của chain chứa node này
        self.pos      = [0]  * n   # vị trí trong segment tree
        self.pos_node = [0]  * n   # node tại vị trí pos (inverse)
        self._cur_pos = 0

    def add_edge(self, u, v):
        self.adj[u].append(v)
        self.adj[v].append(u)

    def build(self):
        self._dfs_size(self.root, -1)
        self._dfs_hld(self.root, self.root)

    def _dfs_size(self, u, p):
        """Tính subtree size + tìm heavy child — iterative để tránh RecursionError."""
        stack = [(u, p, False)]
        order = []
        while stack:
            node, par, done = stack.pop()
            if done:
                for v in self.adj[node]:
                    if v != self.parent[node]:
                        self.subtree[node] += self.subtree[v]
                        if self.heavy[node] == -1 or self.subtree[v] > self.subtree[self.heavy[node]]:
                            self.heavy[node] = v
            else:
                self.parent[node] = par
                stack.append((node, par, True))
                for v in self.adj[node]:
                    if v != par:
                        self.depth[v] = self.depth[node] + 1
                        stack.append((v, node, False))

    def _dfs_hld(self, u, h):
        """Phân rã thành heavy chains — iterative."""
        stack = [(u, h)]
        while stack:
            node, head = stack.pop()
            self.head[node] = head
            self.pos[node] = self._cur_pos
            self.pos_node[self._cur_pos] = node
            self._cur_pos += 1

            # Xử lý light children trước (đẩy vào stack trước)
            for v in self.adj[node]:
                if v != self.parent[node] and v != self.heavy[node]:
                    stack.append((v, v))  # Light child → chain mới

            # Heavy child cuối (LIFO → xử lý tiếp theo)
            if self.heavy[node] != -1:
                stack.append((self.heavy[node], head))

    def lca(self, u, v):
        """LCA sử dụng HLD — O(log N)."""
        while self.head[u] != self.head[v]:
            if self.depth[self.head[u]] < self.depth[self.head[v]]:
                u, v = v, u
            u = self.parent[self.head[u]]
        return u if self.depth[u] < self.depth[v] else v

    def path_query(self, u, v, seg_query):
        """
        Query trên path u→v bằng seg_query(l, r).
        seg_query: function(pos_l, pos_r) → result
        Cần merge result phù hợp với loại query (sum, max, min...).
        """
        result = 0  # Thay identity phù hợp
        while self.head[u] != self.head[v]:
            if self.depth[self.head[u]] < self.depth[self.head[v]]:
                u, v = v, u
            result += seg_query(self.pos[self.head[u]], self.pos[u])   # merge
            u = self.parent[self.head[u]]
        if self.depth[u] > self.depth[v]:
            u, v = v, u
        result += seg_query(self.pos[u], self.pos[v])  # merge
        return result


# ===== USAGE EXAMPLE =====
def solve():
    n, q = map(int, input().split())
    values = list(map(int, input().split()))

    hld = HLD(n, root=0)
    for _ in range(n - 1):
        u, v = map(int, input().split())
        hld.add_edge(u-1, v-1)   # 1-indexed

    hld.build()

    # Tạo Segment Tree theo pos (DFS order của HLD)
    # Đây là nơi plug in SegmentTree từ segment_tree.py
    # from segment_tree import SegmentTree
    # reordered = [values[hld.pos_node[i]] for i in range(n)]
    # st = SegmentTree(n)
    # st.build(reordered)

    # Query
    # for _ in range(q):
    #     u, v = map(int, input().split())
    #     print(hld.path_query(u-1, v-1, st.query))


if __name__ == '__main__':
    solve()

File: skills/cp-solver/references/implementations/hld/hld.cpp
// Heavy-Light Decomposition (HLD) + Segment Tree
// Complexity: O(N log N) preprocessing, O(log² N) per path query/update
// Dùng cho: path queries/updates trên cây

#include <bits/stdc++.h>
using namespace std;

const int MAXN = 100005;
const long long INF = 1e18;

// ===== Segment Tree (range sum, point update) =====
struct SegTree {
  int n;
  vector<long long> tree;
  SegTree(int n) : n(n), tree(4 * n, 0) {}

  void update(int node, int l, int r, int idx, long long val) {
    if (l == r) {
      tree[node] = val;
      return;
    }
    int mid = (l + r) / 2;
    if (idx <= mid)
      update(2 * node, l, mid, idx, val);
    else
      update(2 * node + 1, mid + 1, r, idx, val);
    tree[node] = tree[2 * node] + tree[2 * node + 1];
  }
  long long query(int node, int l, int r, int ql, int qr) {
    if (qr < l || r < ql)
      return 0;
    if (ql <= l && r <= qr)
      return tree[node];
    int mid = (l + r) / 2;
    return query(2 * node, l, mid, ql, qr) +
           query(2 * node + 1, mid + 1, r, ql, qr);
  }
  void update(int idx, long long val) { update(1, 0, n - 1, idx, val); }
  long long query(int l, int r) { return query(1, 0, n - 1, l, r); }
};

// ===== HLD =====
vector<int> adj[MAXN];
int parent[MAXN], depth[MAXN], heavy[MAXN];
int head[MAXN], pos[MAXN], subtree_size[MAXN];
int cur_pos;
int n;

int dfs_size(int u, int p, int d) {
  parent[u] = p;
  depth[u] = d;
  subtree_size[u] = 1;
  int max_sz = 0;
  heavy[u] = -1;
  for (int v : adj[u]) {
    if (v != p) {
      int sz = dfs_size(v, u, d + 1);
      subtree_size[u] += sz;
      if (sz > max_sz) {
        max_sz = sz;
        heavy[u] = v;
      }
    }
  }
  return subtree_size[u];
}

void dfs_hld(int u, int h) {
  head[u] = h;
  pos[u] = cur_pos++;
  if (heavy[u] != -1)
    dfs_hld(heavy[u], h); // Heavy child — cùng chain
  for (int v : adj[u])
    if (v != parent[u] && v != heavy[u])
      dfs_hld(v, v); // Light child — chain mới
}

void hld_build(int root) {
  cur_pos = 0;
  dfs_size(root, -1, 0);
  dfs_hld(root, root);
}

// Path query: sum của các node trên path u→v
long long path_query(int u, int v, SegTree &st) {
  long long result = 0;
  for (; head[u] != head[v]; v = parent[head[v]]) {
    if (depth[head[u]] > depth[head[v]])
      swap(u, v);
    result += st.query(pos[head[v]], pos[v]);
  }
  if (depth[u] > depth[v])
    swap(u, v);
  result += st.query(pos[u], pos[v]);
  return result;
}

// Path update: cập nhật 1 node trên path
void node_update(int u, long long val, SegTree &st) { st.update(pos[u], val); }

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int q;
  cin >> n >> q;

  vector<long long> val(n + 1);
  for (int i = 1; i <= n; i++)
    cin >> val[i];

  for (int i = 0; i < n - 1; i++) {
    int u, v;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
  }

  hld_build(1); // 1-indexed, root = 1

  SegTree st(n);
  for (int i = 1; i <= n; i++)
    st.update(pos[i], val[i]);

  while (q--) {
    int type;
    cin >> type;
    if (type == 1) {
      int u;
      long long v;
      cin >> u >> v;
      node_update(u, v, st);
    } else {
      int u, v;
      cin >> u >> v;
      cout << path_query(u, v, st) << "\n";
    }
  }
  return 0;
}
